### PR TITLE
Add EntityTypeBuilder.BaseType

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Adjust the ModelBuilder logic to account for base types
Discover abstract CLR types

Part of #1704

### CODE REVIEW

File: src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder.cs [hunk 1]
```diff
@@ -307,7 +322,8 @@ protected virtual InternalRelationshipBuilder ReferenceBuilder(EntityType relate
                 navigationToPrincipalName: navigationName ?? "",
                 navigationToDependentName: null,
                 configurationSource: ConfigurationSource.Explicit,
-                strictPrincipal: relatedEntityType == Metadata);
+                strictPrincipal: relatedEntityType.IsAssignableFrom(Metadata)
+                                 || Metadata.IsAssignableFrom(relatedEntityType));
```
Comment (bricelam) [inline src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder.cs]: I don't know what this means, but it seems weird to allow *both* base types and derived types.

File: src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs [hunk 2]
```diff
@@ -49,9 +49,22 @@ public EntityTypeBuilder([NotNull] InternalEntityTypeBuilder builder)
...
+        }
+
+        public new virtual EntityTypeBuilder<TEntity> BaseType([NotNull] Type entityType)
+        {
+            return (EntityTypeBuilder<TEntity>)base.BaseType(entityType);
+        }
+
+        public virtual EntityTypeBuilder<TEntity> BaseType<TBaseType>()
+        {
+            return (EntityTypeBuilder<TEntity>)base.BaseType(typeof(TBaseType));
```
Comment (bricelam) [inline src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs]: Could C# 6-ify these.

File: src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs [hunk 3]
```diff
@@ -19,10 +19,13 @@ public virtual InternalEntityTypeBuilder Apply(InternalEntityTypeBuilder entityT
             Check.NotNull(entityTypeBuilder, nameof(entityTypeBuilder));
             var entityType = entityTypeBuilder.Metadata;
 
-            var keyProperties = DiscoverKeyProperties(entityType);
-            if (keyProperties.Count != 0)
+            if (entityType.BaseType == null)
```
Comment (bricelam) [inline src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs]: Could it be set after this code runs?

File: src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder.cs (hunk 1 above)
Comment (AndriySvyryd) [inline src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder.cs]: You are right, I'll revert this

File: src/EntityFramework7.Core/Internal/PropertyListComparer.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            return result;
+        }
+
+        public bool Equals(IReadOnlyList<IProperty> x, IReadOnlyList<IProperty> y)
+            => Compare(x, y) == 0;
+
+        public int GetHashCode(IReadOnlyList<IProperty> obj)
+            => obj.Aggregate(0, (hash, p) =>
+                {
+                    unchecked
```
Comment (bricelam) [inline src/EntityFramework7.Core/Internal/PropertyListComparer.cs]: Can't we always assumed `unchecked` by default? We're certainly not going to add the compiler flag.

File: src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs]: Resharper should show a warning for this. I no longer read my own code.

File: src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs]: Yes

File: src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs (hunk 2 above)
Comment (anpete) [inline src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs]: It has the quick-fix. You can probably change the severity.

File: src/EntityFramework7.Core/Internal/PropertyListComparer.cs (hunk 4 above)
Comment (AndriySvyryd) [inline src/EntityFramework7.Core/Internal/PropertyListComparer.cs]: I don't see why we wouldn't do it at some point for a debug build.
I'll use the lambda form:
```
obj.Aggregate(0, (hash, p) => unchecked((hash * 397) ^ p.GetHashCode()))
```

File: test/EntityFramework7.Core.FunctionalTests/TestUtilities/PropertyComparer.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+            return x.Name == y.Name
+                   && x.ClrType == y.ClrType
+                   && x.IsNullable == y.IsNullable
+                   && x.IsConcurrencyToken == y.IsConcurrencyToken
+                   && x.RequiresValueGenerator == y.RequiresValueGenerator
+                   && x.ValueGenerated == y.ValueGenerated
+                   && x.IsReadOnlyBeforeSave == y.IsReadOnlyBeforeSave
+                   && x.IsReadOnlyAfterSave == y.IsReadOnlyAfterSave
+                   && x.SentinelValue == y.SentinelValue;
```
Comment (bricelam) [inline test/EntityFramework7.Core.FunctionalTests/TestUtilities/PropertyComparer.cs]: Should this (and other comparers) compare annotations?

Comment (AndriySvyryd) [inline test/EntityFramework7.Core.FunctionalTests/TestUtilities/PropertyComparer.cs]: Maybe, depends on what you are testing

File: src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs (hunk 3 above)
Comment (bricelam) [inline src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs]: Should existing keys be removed in that case?

**Comment by bricelam** [comment]
:shipit:

File: src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EntityFramework7.Core/Metadata/Builders/EntityTypeBuilder`.cs]: Found it. Will include in the commit.

File: src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs (hunk 3 above)
Comment (AndriySvyryd) [inline src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs]: Setting BaseType would have already done that

Comment (bricelam) [inline src/EntityFramework7.Core/Metadata/Conventions/Internal/KeyDiscoveryConvention.cs]: Now that you mention it, I do remember seeing that code... :grinning:

**Comment by smitpatel** [comment]
Committed to release, merge to dev?

**Pull request merged**